post arc review modifications (#55)
riscv-non-isa/riscv-security-model: specification/src/chapter3.adoc

| SR_PMP_001
| PMP configurations MUST only be directly accessible to machine mode
| SR_PMP_002

> | PMP/smepmp MUST be used to isolate M-mode from lower privilege levels

Reviewer: should we make this allow more options if its a MUST ? 

Either PMP/Smepmp OR MPT MUST be used to isolate M-mode from lower privilege levels

Author: yes, agree

> | PMP/smepmp MUST be used to isolate M-mode from lower privilege levels

Author: as per Ravi:
```suggestion
| PMP/smepmp OR MPT MUST be used to isolate M-mode from lower privilege levels
```

| ID#
| Requirement

| SR_PMP_004

> | sPMP MUST be used to protect S-mode from lower privilege levels.

Reviewer: I liked the previous wording which provided some flexibility - 

if MMU is not supported, sPMP SHOULD be used to protect S-Mode from lower priv levels

Author: i think this is one where there are lots of possible options - even with MMU supported there may be cases where sPMP is mandated to protect S-mode, hence this as a referenceable SR.  do we need a note to point out that this SR would only be mandated in certain circumstances ?

Reviewer: the way it is written right  now - it reads like an absolute - if it is an optional, it should be written as a SHOULD - how about:

sPMP or MMU (Sv) MUST be used to protect S-mode from lower privilege levels. Even with MMU (Sv) implemented, sPMP MAY still be used in addition.

Author: ```suggestion
sPMP or MMU (Sv) MUST be used to protect S-mode from lower privilege levels. Even with MMU (Sv) implemented, sPMP MAY still be used in addition.| sPMP MUST be used to protect S-mode from lower privilege levels.
```

| SR_PMP_005

> | sPMP MUST be used to protect resources assigned to one workload from other workloads

Reviewer: how is this different from 004?

Author: sPMP used by OS to create isolated U mode domains ?

Author: ```suggestion
| sPMP or MMU (Sv) MUST be used to protect U-mode workloads from other U-mode workloads. Even with MMU (Sv) implemented, sPMP MAY still be used in addition for U-mode inter-workload isolation.
```

| Requirement

| SR_MMU_001
| if the Sv extension is supported then 1st-stage page tables MUST used to protect
the S-Mode Supervisor domain from accesses made by U-Mode.

| SR_MMU_002
| if the H extension is supported 1st-stage and/or G-stage page tables MUST used to protect
Supervisor domain H/S-mode from lower privilege levels.

| SR_MMU_003

> | MMU MUST be used to protect resources assigned to one workload from other workloads

Reviewer: should clarify - the workload here is guest virtual machines i presume - then should specify use G-stage page tables to protect VM workloads from other VM workloads.

Reviewer 2: My reading is that the workload in reference can be under a hypervisor or not. See line 169

Author: There will be systems where this is the required means to isolate (under hypervisor or not), and other systems where MPT could be used. I'd expect both to be listed as MUST SRs,  then specific use cases to reference, one or other, or allow either as an option.

Reviewer: ok - the presentation of all these options is something we should cleanup past this next ARC review. I have some thoughts on how to summarize.

|===

> NOTE: Dependent on the use case, it may be recommended or mandatory to protect supervisor mode from lower privilege levels (with or without the H-extension). Trusted execution environments may require the use of MMU to isolate trusted applications from each other. The sPMP may be an additional or alternative mechanism.

Reviewer: ```suggestion
NOTE: Dependent on the use case, it may be recommended or mandatory to protect supervisor mode from lower privilege levels (with or without the H-extension). Trusted execution environments may require the use of MMU to isolate trusted applications from each other. The sPMP and PMP may be an additional or alternative mechanism.
```

==== Supervisor domains and MPT

*See the https://github.com/riscv/riscv-smmtt[RISC-V Supervisor Domains Access
Protection] specification.*

Supervisor domains allow software components on the same hart to be developed,
certified, deployed and attested independently of each other.

A supervisor domain is an S-Mode compartment that is physically isolated from other supervisor domains. The memory,
execution state and devices belonging to a supervisor domains are isolated from other supervisor domains.
This isolation of supervisor domains and the context switching between them is managed by M-mode firmware.

> A supervisor domain is identified at an architecture level by a _supervisor domain
> id (SDID)_ CSR, managed by M-mode firmware.

Reviewer: ```suggestion
A supervisor domain is identified at an architecture level by a _supervisor domain
id (SDID)_ held in the _mmpt_ CSR, managed by M-mode firmware. In addition to
the SDID, the mmpt CSR may specify a root PPN for a Memory Protection Table (MPT).
```

Reviewer 2: This is a good point that completely went over my head, the SDID is not global across different MPTs.

Reviewer 3: I think SDID is global.

PPN can have a lot of bits, and SDID+PPN identifier would force implementations to carry around a lot of mostly useless bits for all memory operations.

> The _memory protection table (MPT)_ is a memory structure managed by machine
> mode that is used to track memory ownership across supervisor domains. It is designed to
> enable fine grained dynamic memory management across supervisor domain
> boundaries, with policy typically set by a hypervisor in a hosting domain

Reviewer: ```suggestion
The _memory protection table (MPT)_ is a memory structure managed by machine
mode that is used to manage physical memory permissions across supervisor domains.
It is designed to enable page-based dynamic memory management across supervisor
domain boundaries.
```

| ID#
| Requirement

| SR_MPT_001
| MPT MUST be used to protect M-mode from lower privilege
levels

| SR_MPT_002

> | MPT configurations MUST only be directly accessible to machine mode

Reviewer: ```suggestion
| MPT configurations SHOULD only be directly accessible to machine mode.
```

Reviewer 2: Out of curiosity, why are you proposing to weaken the wording here? What is the usecase where the MPT configurations are modified by lesser privileged levels?

Author: are you imagining one supervisor domain that can control the MPT configurations ?  else it doesnt make sense to me

Author: right, makes sense.  need to add a note justifying

|===

NOTE: The M-Mode resident software responsible for managing context switches and communication between supervisor

> domains is called the Root Domain. Depending on the use case, MTT can be sufficient for protecting the Root Domain by enabling M-mode

Reviewer: ```suggestion
domains is called the Root Domain. Depending on the use case, MPT can
be sufficient for protecting the Root Domain by enabling M-mode
```

| Requirement

| SR_SUD_001
| PMP/Smepmp or MPT MUST be used to enforce physical memory isolation
boundaries for supervisor domains, and to protect machine mode from any
supervisor domain.

|===

PMP can be used for more static and deterministic use cases.

> MPT can be used where more fine grained dynamic resource management across

Reviewer: ```suggestion
MPT can be used where more fine grained (page-based) dynamic resource management across
```

supervisor domain boundaries is required.

===== Supervisor Domain Resource Access and Sharing

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| SR_SUD_002
| MPT MUST be used to protect resources assigned to one supervisor domain from other supervisor domains

| SR_SUD_003

> | Resources assigned to a trusted supervisor domain MUST be inaccessible to an untrusted supervisor domain

Reviewer: ```suggestion
| Resources assigned to a higher trust level supervisor domain MUST be inaccessible
to an lower trust level supervisor domain. The trust levels/policies are specified by the
security system designer.
```

| SR_SUD_004

> | Resources assigned to an untrusted supervisor domain MUST be accessible to a trusted supervisor domain 

Reviewer: should we restate these as:
Resources assigned to a supervisor domain MUST be accessible to another supervisor domain only under the control of the higher privilege root domain supervisor via MPT permissions.

Author: i added these to allow the GP TEE example section to directly reference as requirements - it expects this asymmetric sharing model. Your modification encompasses the RVI requirement. So should be  reword the example section and modify as you suggest ?

Reviewer: issue is stating trusted and untrusted is subjective. see my suggestion for the previous SR as well. 

how about:
Resources assigned to a lower trust level supervisor domain MAY be accessible to a higher trust level supervisor domain. The trust levels/policies are specified by the security system designer.

Reviewer 2: This comment goes hand-in-hand with the discussions we had with @yloisel in another commit. I agree with Ravi's proposal.

Author: I agree stating trusted and untrusted is subjective in the general case - but there are models where this is strictly defined (eg. GP TEE).  how about these two are replaced with 
| SR_SUD_004
|  Access to resources assigned to a supervisor domain from other supervisor domains MUST be enforced according to a defined policy

|===

> Supervisor domains allow resource isolation and sharing between domains under the control of M-mode firmware. Trusted Execution environments can require asymmetric sharing models, where one trusted domain has R/W access to other domain's resources.

Reviewer: ```suggestion
Supervisor domains allow resource isolation and sharing between domains under the control of M-mode firmware. Trusted Execution environments can require asymmetric sharing models, where one trusted domain has X/W/R access to other domain's resources.
```

extensions for external debug (TBD).

|===

*See chapter 6 (Smsdia) of the https://github.com/riscv/riscv-smmtt[RISC-V Supervisor
Domains Access Protection] specification.*

==== External debug and Performance counters

*See the https://github.com/riscv-non-isa/riscv-external-debug-security[RISC-V External Debug Security Extension]
specification.*

[cat_sr_sub_dbg]
[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| SR_DBG_001
| External debug MUST only be enabled by HW RoT (M-mode external debug) or by FW
RoT (non M-mode external debug).

| SR_DBG_002
| External debug SHOULD be enabled separately for M-mode & non-M-mode software.

| SR_DBG_003
| Self-hosted debug MAY be used for debug of non M-mode software.

| SR_DBG_004
| Self-hosted debug MUST only be enabled by a higher privileged component.

|===

For example, external debug can be enabled for non-M-mode software without affecting M-mode (recoverable debug). And an S-mode OS can enable self-hosted debug for a user application without affecting other applications or S-mode itself.

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| SR_DBG_005

> | FW RoT MAY disable self-hosted debug for all non M-mode software.

Reviewer: should this say RoT instead of FW RoT - also why does this refer to non M-mode sw only?

Author: i think this is for the scenario m-mode is controlling self hosted debug for lower privilege SW

Reviewer: then we should say something like M-mode MAY extend the HW RoT to implement a FW RoT. and seperately state the FW RoT can do what this SR suggests.

Author: do the existing RoT definitions in Ch2 already cover this ?

|===

For example, disable self-hosted debug in a production system for certification
reasons.

[width=100%]
[%header, cols="5,20"]
|===
| ID#
| Requirement

| SR_DBG_006
| External debug MUST only be enabled following system reset (part of measuring)
of the affected component, moderated by a RoT.

| SR_DBG_007
| Revealing self-hosted debug MUST only be enabled following reboot (part of
measuring) of the affected component.

| SR_DBG_008
| Trusted self-hosted debug MAY be enabled at runtime (after measuring) of the
affected component, to an application specific governance process.

|===

> Guarantees the system remains attestable.

Reviewer: straggler text? should this be anchored in DBG_008?

Author: it refers to the SRs in the table above - maybe it should be reworded

```suggestion
These SRs are to ensure that the system remains attestable.
```

> | A system which supports PMP/Smepmp, or sPMP, MUST implement IOPMP for device
> access control unless the system supports IOMTT.

Reviewer: ```suggestion
| A system which supports PMP/Smepmp, or sPMP, MUST implement either IOPMP or IOMPT for device
access control.
```

> Depending on system design, IOMTT can enforce the same access control policies as IOPMP.

Reviewer: ```suggestion
Depending on system design, IOMPT can enforce the same access control policies as IOPMP.
```

| SR_IOP_002
| IOPMP configurations MUST only be directly accessible to machine mode.

| SR_IOP_003
| IOPMP MUST be used to guarantee that devices assigned to lower privilege levels cannot access resources assigned to M-mode.

| SR_IOP_004
| IOPMP MUST be used to guarantee that devices assigned to a domain cannot be accessed by other domains.

|===

NOTE: IOPMP defines multiple "models" for different system configurations.
Unless specified differently in the use cases in this specification, system
designers are free to choose any IOPMP model. Depending on the use case, IOMPT may be an alternative or addition to IOPMP

==== IOMPT

*See the https://github.com/riscv/riscv-smmtt[RISC-V Supervisor Domains Access
Protection] specification.*

IOMPT is a system level component providing physical memory access control for

> device-initiated transactions, complementing MTT rules.

Reviewer: ```suggestion
device-initiated transactions, by mapping transactions to IOMMU translation and MPT permissions.
```

| ID#
| Requirement

| SR_IOM_006
| Systems supporting MMU SHOULD also support IOMMU

| SR_IOM_007
| Systems supporting IOMMU MUST also enforce physical memory access control for

> M-mode memory against device-initiated transactions using IOMTT or IOPMP

Reviewer: ```suggestion
M-mode memory against device-initiated transactions using IOMPT or IOPMP
```
